In [114]:
import numpy as np
import pandas as pd
from datetime import datetime

In [115]:
raw_data = pd.read_csv('train.csv')

In [116]:
raw_data.head()

Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

In [117]:
raw_data.describe(include='all')

Unnamed: 0                    Name Location         Year  \
count   5847.000000                    5847     5847  5847.000000   
unique          NaN                    1804       11          NaN   
top             NaN  Mahindra XUV500 W8 2WD   Mumbai          NaN   
freq            NaN                      49      762          NaN   
mean    3013.181461                     NaN      NaN  2013.448435   
std     1736.398890                     NaN      NaN     3.194949   
min        1.000000                     NaN      NaN  1998.000000   
25%     1509.500000                     NaN      NaN  2012.000000   
50%     3015.000000                     NaN      NaN  2014.000000   
75%     4517.500000                     NaN      NaN  2016.000000   
max     6018.000000                     NaN      NaN  2019.000000   

        Kilometers_Driven Fuel_Type Transmission Owner_Type    Mileage  \
count        5.847000e+03      5847         5847       5847       5845   
unique                NaN         3            2          4        419   
top                   NaN    Diesel       Manual      First  18.9 kmpl   
freq                  NaN      3161         4135       4811        172   
mean         5.841013e+04       NaN          NaN        NaN        NaN   
std          9.237971e+04       NaN          NaN        NaN        NaN   
min          1.710000e+02       NaN          NaN        NaN        NaN   
25%          3.346750e+04       NaN          NaN        NaN        NaN   
50%          5.257600e+04       NaN          NaN        NaN        NaN   
75%          7.249050e+04       NaN          NaN        NaN        NaN   
max          6.500000e+06       NaN          NaN        NaN        NaN   

         Engine   Power        Seats   New_Price        Price  
count      5811    5811  5809.000000         815  5847.000000  
unique      140     361          NaN         534          NaN  
top     1197 CC  74 bhp          NaN  95.13 Lakh          NaN  
freq        602     235          NaN           6          NaN  
mean        NaN     NaN     5.286452         NaN     9.653742  
std         NaN     NaN     0.806668         NaN    11.275966  
min         NaN     NaN     2.000000         NaN     0.440000  
25%         NaN     NaN     5.000000         NaN     3.550000  
50%         NaN     NaN     5.000000         NaN     5.750000  
75%         NaN     NaN     5.000000         NaN    10.250000  
max         NaN     NaN    10.000000         NaN   160.000000

**a) Look for the missing values in all the columns and either impute them (replace with mean,
median, or mode) or drop them. Justify your action for this task.**

In [118]:
missing_values = raw_data.isnull().sum()
print("Missing Values:")
print(missing_values)

Missing Values:
Unnamed: 0              0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  38
New_Price            5032
Price                   0
dtype: int64


Question b need to be addressed before completing question a

**b) Remove the units from some of the attributes and only keep the numerical values (for
example remove kmpl from “Mileage”, CC from “Engine”, bhp from “Power”, and lakh from
“New_price”).**

In [119]:
raw_data['Mileage'] = pd.to_numeric(raw_data['Mileage'].str.replace(' km/kg| kmpl', '', regex=True), errors='coerce')
raw_data['Engine'] = pd.to_numeric(raw_data['Engine'].str.replace(' CC', '', regex=True), errors='coerce')
raw_data['Power'] = pd.to_numeric(raw_data['Power'].str.replace(' bhp', '', regex=True), errors='coerce')
raw_data['New_Price'] = pd.to_numeric(raw_data['New_Price'].str.replace(' Lakh', '', regex=True), errors='coerce')


In [120]:
print("\nDataFrame after removing units:")
print(raw_data)


DataFrame after removing units:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0                 41000 

**Question A continuation**

In [121]:
# Impute missing values
raw_data['Mileage'].fillna(raw_data['Mileage'].mean(), inplace=True)
raw_data['Engine'].fillna(raw_data['Engine'].mean(), inplace=True)
raw_data['Power'].fillna(raw_data['Power'].mean(), inplace=True)
raw_data['Seats'].fillna(raw_data['Seats'].mode()[0], inplace=True)

In [122]:
raw_data.drop('New_Price', axis=1, inplace=True)

a. Justification:

I have used mean for numerical data and mode for categorical data when dealing with small number of missing values. The 'New_Price' column had a significant number of missing values (5032 out of the total rows), making imputation less reliable. Dropping it avoids potential bias.

In [123]:
print("\nDataFrame after handling missing values:")
print(raw_data)


DataFrame after handling missing values:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Fuel_Type Transmission Owner_Type  Mileage  Engine  \
0              

**C) Change the categorical variables (“Fuel_Type” and “Transmission”) into numerical one hot
encoded value**

In [124]:
df = pd.get_dummies(raw_data, columns=['Fuel_Type', 'Transmission'], drop_first=True)


In [125]:
print("\nDataFrame after one-hot encoding categorical variables:")
print(df)


DataFrame after one-hot encoding categorical variables:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0

**d) Create one more feature and add this column to the dataset**

In [126]:
current_year = datetime.now().year
df['Current_Age'] = current_year - df['Year']

In [127]:
print("\nDataFrame with the new 'Current_Age' feature:")
print(df)


DataFrame with the new 'Current_Age' feature:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kilometers_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0          

**e) Perform select, filter, rename, mutate, arrange and summarize with group by operations (or
their equivalent operations in python) on this dataset.**

Select Operation:

In [128]:
selected_columns = df[[ 'Owner_Type', 'Price']]
print("\nSelected Columns:")
print(selected_columns)


Selected Columns:
     Owner_Type  Price
0         First  12.50
1         First   4.50
2         First   6.00
3        Second  17.74
4         First   3.50
...         ...    ...
5842      First   4.75
5843      First   4.00
5844     Second   2.90
5845      First   2.65
5846      First   2.50

[5847 rows x 2 columns]


Filter Operation:

In [129]:
filtered_data = df[df['Location'] == 'Chennai']
print("\nFiltered Data (Diesel cars only):")
print(filtered_data)


Filtered Data (Diesel cars only):
      Unnamed: 0                                    Name Location  Year  \
1              2                            Honda Jazz V  Chennai  2011   
2              3                       Maruti Ertiga VDI  Chennai  2012   
7              9          Tata Indica Vista Quadrajet LS  Chennai  2012   
37            39   Volkswagen Polo Diesel Trendline 1.2L  Chennai  2013   
52            54                Hyundai Grand i10 Sportz  Chennai  2015   
...          ...                                     ...      ...   ...   
5762        5930         Maruti Swift Dzire VXI Optional  Chennai  2015   
5764        5932           Maruti Alto 800 2016-2019 LXI  Chennai  2016   
5784        5954                  Toyota Fortuner 4x4 AT  Chennai  2015   
5786        5956  Toyota Innova 2.5 GX (Diesel) 7 Seater  Chennai  2012   
5827        5999                   Tata Bolt Revotron XT  Chennai  2016   

      Kilometers_Driven Owner_Type  Mileage  Engine   Power  Sea

Rename Operation:

In [130]:
df = df.rename(columns={'Kilometers_Driven': 'Kms_Driven'})
print("\nDataFrame with Renamed Columns:")
print(df)


DataFrame with Renamed Columns:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kms_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0          41000      First    1

Mutate operation:

In [131]:
df['Power_Per_Seat'] = df['Power'] / df['Seats']
print("\nDataFrame with New 'Power_Per_Seat' Column:")
print(df)


DataFrame with New 'Power_Per_Seat' Column:
      Unnamed: 0                              Name    Location  Year  \
0              1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1              2                      Honda Jazz V     Chennai  2011   
2              3                 Maruti Ertiga VDI     Chennai  2012   
3              4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4              6            Nissan Micra Diesel XV      Jaipur  2013   
...          ...                               ...         ...   ...   
5842        6014                  Maruti Swift VDI       Delhi  2014   
5843        6015          Hyundai Xcent 1.1 CRDi S      Jaipur  2015   
5844        6016             Mahindra Xylo D4 BSIV      Jaipur  2012   
5845        6017                Maruti Wagon R VXI     Kolkata  2013   
5846        6018             Chevrolet Beat Diesel   Hyderabad  2011   

      Kms_Driven Owner_Type  Mileage  Engine   Power  Seats  Price  \
0          41000    

Arrange operation:

In [132]:
arranged_data = df.sort_values(by='Year', ascending=False)
print("\nDataFrame Sorted by 'Year' in Descending Order:")
print(arranged_data)


DataFrame Sorted by 'Year' in Descending Order:
      Unnamed: 0                               Name    Location  Year  \
5405        5560          Renault KWID RXT Optional       Kochi  2019   
942          975    Ford Endeavour 2.2 Trend AT 4X2       Kochi  2019   
5533        5690          Maruti Omni 5 Seater BSIV  Coimbatore  2019   
770          796              Mahindra XUV500 W9 AT  Coimbatore  2019   
4267        4399         Maruti Swift Dzire AMT ZDI     Chennai  2019   
...          ...                                ...         ...   ...   
1185        1224                      Maruti Zen VX      Jaipur  1999   
1791        1845                 Honda City 1.3 EXI        Pune  1999   
3630        3749  Mercedes-Benz E-Class 250 D W 210      Mumbai  1998   
5558        5716                      Maruti Zen LX      Jaipur  1998   
3039        3138                     Maruti Zen LXI      Jaipur  1998   

      Kms_Driven Owner_Type  Mileage  Engine  Power  Seats  Price  \
5405 

In [133]:
summary_data = df.groupby('Fuel_Type_Electric')['Power'].mean().reset_index()
print("\nMean Price for Electric Fuel Type:")
print(summary_data)


Mean Price for Electric Fuel Type:
   Fuel_Type_Electric       Power
0                   0  113.822581
1                   1   57.000000


In [134]:
df.to_csv("train_final.csv", index=False)